<?xml version="1.0"?>
<robot name="firstRobot" xmlns:xacro="http://www.ros.org/wiki/xacro">
    <xacro:property name="params" value="${load_yaml('$(find lab5)/urdf_parameters.yaml')}"/>
    <xacro:property name="l_width" value="0.1"/>
 
    <material name="red">
        <color rgba="0.80392157 0 0 1"/>
    </material>

    <material name="blue">
        <color rgba="0 0.1 0.7 1"/>
    </material>

    <material name="black"> 
        <color rgba="0.05 0.05 0.05 1"/>
    </material>

    <link name="base_link">
        <visual>
            <geometry>
                <cylinder length="0.1" radius="0.3"/>
            </geometry>
            <origin xyz="0 0 -0.05"/>
            <material name="black"/>
        </visual>
    </link>

    <link name="link_1">
        <visual>
            <geometry>
                <box size="${l_width} ${l_width} ${params['i2']['l_len']}"/>
            </geometry>
            <origin xyz="${params['i2']['l_xyz']}" rpy="0 0 0"/>
            <material name="blue"/>
        </visual>
    </link>

    <link name="link_2">
        <visual>
            <geometry>
                <box size="${l_width} ${l_width} ${params['i3']['l_len']}"/>
            </geometry>
            <origin xyz="${params['i3']['l_xyz']}" rpy="0 1.57079632679 0"/>
            <material name="blue"/>
        </visual>
    </link>

    <link name="link_3">
        <visual>
            <geometry>
                <box size="${l_width} ${l_width} ${params['hand']['l_len']}"/>
            </geometry>
            <origin xyz="${params['hand']['l_xyz']}" rpy="0 1.57079632679 0"/>
            <material name="blue"/>
        </visual>
    </link>

    <link name="hand">
        <visual>
            <geometry>
                <box size="${l_width} ${l_width} ${l_width}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <material name="red"/>
        </visual>
    </link>       

    <joint name="base_to_link1" type="fixed">
        <parent link="base_link"/>
        <child link="link_1"/>
        <origin xyz="${params['i1']['j_xyz']}" rpy="${params['i1']['j_rpy']}"/>
    </joint>

    <joint name="link1_to_link2" type="fixed">
        <parent link="link_1"/>
        <child link="link_2"/>
        <origin xyz="${params['i2']['j_xyz']}" rpy="${params['i2']['j_rpy']}"/>
    </joint>

    <joint name="link2_to_link3" type="fixed">
        <parent link="link_2"/>
        <child link="link_3"/>
        <origin xyz="${params['i3']['j_xyz']}" rpy="${params['i3']['j_rpy']}"/>
    </joint>

    <joint name="link3_to_hand" type="fixed">
        <parent link="link_3"/>
        <child link="hand"/>
        <origin xyz="${params['hand']['j_xyz']}" rpy="${params['hand']['j_rpy']}"/>
    </joint>

</robot>
